crates/bootstrap_srv/src/test.rs

```diff
@@ -0,0 +1,654 @@
+        info: info1,
+        agent: agent1,
+    } = PutInfo {
+        addr: s.listen_addr(),
+        ..Default::default()
+    }
+    .call()
+    .unwrap();
+
+    // -- put agent2 -- //
+
+    let PutInfoRes {
+        info: info2,
+        agent: agent2,
+    } = PutInfo {
+        addr: s.listen_addr(),
+        agent_seed: K2,
+        ..Default::default()
+    }
+    .call()
+    .unwrap();
+
+    // -- tombstone agent1 -- //
+
+    let PutInfoRes {
+        info: info1_t,
+        agent: agent1_t,
+    } = PutInfo {
+        addr: s.listen_addr(),
+        is_tombstone: true,
+        ..Default::default()
+    }
+    .call()
+    .unwrap();
+
+    // -- validate test -- //
+
+    assert_eq!(agent1, agent1_t);
+    assert_ne!(agent1, agent2);
+    assert_ne!(info1, info2);
+    assert_ne!(info1, info1_t);
+
+    // -- get the result -- //
+
+    let addr = format!("http://{:?}/bootstrap/{}", s.listen_addr(), S1);
+    let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+    let mut res: Vec<DecodeAgent> = serde_json::from_str(&res).unwrap();
+
+    assert_eq!(1, res.len());
+    let one = res.pop().unwrap();
+    assert_eq!(one.agent, agent2);
+}
+
+#[test]
+fn reject_mismatch_agent_url() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let err = PutInfo {
+        addr: s.listen_addr(),
+        agent_url: Some("AAAA"),
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidAgent"));
+}
+
+#[test]
+fn reject_mismatch_space_url() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let err = PutInfo {
+        addr: s.listen_addr(),
+        space_url: "AAAA",
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidSpace"));
+}
+
+#[test]
+fn reject_old_created_at() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let err = PutInfo {
+        addr: s.listen_addr(),
+        created_at: 0,
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidCreatedAt"));
+}
+
+#[test]
+fn reject_future_created_at() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let err = PutInfo {
+        addr: s.listen_addr(),
+        created_at: i64::MAX - 500,
+        expires_at: i64::MAX,
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidCreatedAt"));
+}
+
+#[test]
+fn reject_expired() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let expires_at = crate::now() - 500;
+    let created_at = crate::now() - 1500;
+
+    let err = PutInfo {
+        addr: s.listen_addr(),
+        created_at,
+        expires_at,
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidExpiresAt"));
+}
+
+#[test]
+fn reject_expired_at_before_created_at() {
```

Comment:
Maybe another test for rejecting created_at == expired_at

Reply:
Done: https://github.com/holochain/kitsune2/pull/11/commits/8cfe688c6779c30e6b214b4bbcfb08eaa3da4652